zeq/publisher.cpp

```diff
@@ -16,21 +17,55 @@
 #  include <zerobuf/Zerobuf.h>
 #endif
 
+#include <cstring>
 #include <map>
-#include <string.h>
 
 namespace zeq
 {
-namespace detail
+
+namespace
+{
+std::string _getApplicationName()
 {
-class Publisher : public Sender
+    // http://stackoverflow.com/questions/933850
+#ifdef _MSC_VER
+    char result[MAX_PATH];
+    const std::string execPath( result, GetModuleFileName( NULL, result,
+                                                           MAX_PATH ));
+#elif __APPLE__
+    char result[PATH_MAX+1];
+    uint32_t size = sizeof(result);
+    if( _NSGetExecutablePath( result, &size ) != 0 )
+        return std::string();
+    const std::string execPath( result );
+#else
+    char result[PATH_MAX];
+    const ssize_t count = readlink( "/proc/self/exe", result, PATH_MAX );
+    const std::string execPath( result, count > 0 ? count : 0 );
+#endif
+
+    size_t lastSeparator = 0;
+    const size_t length = execPath.length();
+
+    for( size_t i = 0; i < length; ++i )
+        if( execPath[ i ] == '/' || execPath[i] == '\\' )
+            lastSeparator = i+1;
+
+    return lastSeparator == 0 ? execPath :
+                                execPath.substr( lastSeparator, length );
```

Comment:
substr takes position and count not two positions, this will access out-of-bounds memory, use substr( lastSeparator ), that will go until the end.